docs/developer/best_practices/commit.rst

```diff
@@ -0,0 +1,356 @@
+Commit best practices
+---------------------
+
+How to split a change into commits
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The golden rule to create good commits is to ensure that there is only one
+logical change per commit.
+
+This has several advantages amongst whose:
+
+- small commits are easier to review (a large pull request correctly divided
+  into logical commits is easier/faster to review than a medium-sized one with
+  less thought-out division)
+
+- logical commits are easier to revert (`e866b01f0553 <https://github.com/scality/metalk8s/commit/e866b01f05535925e80da20aca00417904422433>`_/`8208a170ac66 <https://github.com/scality/metalk8s/commit/8208a170ac66912ace018bcd00c058ad214d169b>`_)/cherry-pick
+  (`Pull request #1641 <https://github.com/scality/metalk8s/pull/1641>`_)
+
+- when looking for a regression (e.g. using ``git bisect``) it is easier to
+  find the root cause
+
+- make ``git log`` and ``git blame`` way more useful
+
+Examples
+^^^^^^^^
+
+Cosmetic changes
+::::::::::::::::
+
+Use a dedicated commit when you want to make cosmetic changes to the code
+(linting, whitespaces, alignment, renaming, …).
+
+Mixing cosmetics and functional changes is bad because the cosmetics (which
+tend to generate a lot of diff/noise) will obscure the important functional
+changes, making it harder to correctly determine whether the change is correct
+during the review.
+
+Example (Pull request `#1620 <https://github.com/scality/metalk8s/issues/1620>`_):
+
+- one commit for the cosmetic changes: `766f572e462c6933c8168a629ed4f479bb68a803 <https://github.com/scality/metalk8s/commit/766f572e462c6933c8168a629ed4f479bb68a803>`_
+
+- one commit for the functional changes: `3367fabdefc0b35d34bf7cf2fb0d33ff81f9fd5a <https://github.com/scality/metalk8s/commit/3367fabdefc0b35d34bf7cf2fb0d33ff81f9fd5a>`_
+
+Refactoring
+:::::::::::
+
+When introducing new features, you often have to add new helpers or refactor
+existing code. In such case, instead of having single commit with everything
+inside, you can either:
+
+1. first add a new helper: `29f49cbe9dfa <https://github.com/scality/metalk8s/commit/29f49cbe9dfa0b824c818d25d4a2f6965351e65d>`_
+
+2. then use it in new code: `7e47310a8f20 <https://github.com/scality/metalk8s/commit/7e47310a8f20fd49f0ad36707b20e6c2a53df638>`_
+
+Or:
+
+1. first add the new code: `5b2a6d5fa498 <https://github.com/scality/metalk8s/commit/5b2a6d5fa49815180a2effdd37cb58542e83b5a5>`_
```

Comment:
Again, I'm not sure doing this in 2 commits brings any value. It'd require *both* commits to be reverted and/or cherry-picked to 'work'.

Reply:
Personally, it is easier for me to review if they are in one commit since I would like know why is it refactored and how is it used

Reply:
Now, if this example is really controversial/confusing I can remove it.

It's no big deal, given that:
- people usually split their changes into commit already (most of the time at least), our current biggest problem are the commit messages…
- those examples are only a guidance at best, not something to strictly enforce.

Opinion @NicolasT , @carlito-scality ?